frontendActions/src/main/java/com/virtuslab/gitmachete/frontend/actions/common/PullBackgroundable.java

```diff
@@ -0,0 +1,214 @@
+          AbstractCommonUpdateAction
+              .showsCustomNotification(java.util.Collections.singletonList(GitVcs.getInstance(project)))) {
+        var ranges = updatedRanges.calcCurrentPositions();
+        GitUpdateInfoAsLog.NotificationData notificationData = new GitUpdateInfoAsLog(project, ranges)
+            .calculateDataAndCreateLogTab();
+
+        Notification notification;
+        if (notificationData != null) {
+          String title = getTitleForUpdateNotification(notificationData.getUpdatedFilesCount(),
+              notificationData.getReceivedCommitsCount());
+          String content = getBodyForUpdateNotification(notificationData.getFilteredCommitsCount());
+          notification = VcsNotifier.STANDARD_NOTIFICATION.createNotification(title,
+              content,
+              INFORMATION,
+              /* listener */ null);
+          notification.addAction(NotificationAction.createSimple(getString(
+              "action.GitMachete.PullBackgroundable.notification.message.view-commits"),
+              notificationData.getViewCommitAction()));
+
+        } else {
+          // When the pull results with no commits, there is no git update info (as log).
+          // Based on that we know that all files are up-to-date.
+          notification = VcsNotifier.STANDARD_NOTIFICATION.createNotification(
+              getString("action.GitMachete.PullBackgroundable.notification.title.all-files-are-up-to-date"),
+              /* content */ "", INFORMATION, /* listener */ null);
+        }
+        VcsNotifier.getInstance(project).notify(notification);
+
+      } else {
+        showUpdates(project, root, currentRev, beforeLabel);
+      }
+
+    } else if (localChangesDetector.wasMessageDetected()) {
+      LocalChangesWouldBeOverwrittenHelper.showErrorNotification(project,
+          repository.getRoot(),
+          OPERATION_NAME,
+          localChangesDetector.getRelativeFilePaths());
+
+    } else if (untrackedFilesDetector.wasMessageDetected()) {
+      GitUntrackedFilesHelper.notifyUntrackedFilesOverwrittenBy(project,
+          root,
+          untrackedFilesDetector.getRelativeFilePaths(),
+          OPERATION_NAME,
+          /* description */ null);
+
+    } else {
+      GitUIUtil.notifyError(project,
+          getString("action.GitMachete.PullBackgroundable.notification.fail"),
+          result.getErrorOutputAsJoinedString(),
+          /* important */ true,
+          /* error */ null);
+      repository.update();
+    }
+  }
+
+  private static void showUpdates(Project project, VirtualFile root, GitRevisionNumber currentRev, Label beforeLabel) {
+    try {
+      UpdatedFiles files = UpdatedFiles.create();
+      MergeChangeCollector collector = new MergeChangeCollector(project, root, currentRev);
+      collector.collect(files);
+
+      GuiUtils.invokeLaterIfNeeded(() -> {
+        ProjectLevelVcsManagerEx manager = (ProjectLevelVcsManagerEx) ProjectLevelVcsManager.getInstance(project);
```

Comment:
Maybe drop a comment about why can be sure that this instance is a `ProjectLevelVcsManagerEx`

Reply:
"They" do it too